# **Aryan_MDM_NLP**
## Objective: Implement NLP preprocessing and build a text classification model.

### **Install Requirements**

In [9]:
!pip install pandas numpy nltk scikit-learn

### **1. Import Libraries and NLP Preprocessing**
- Tokenization
- Stopword Removal
- Stemming
- Lemmatization

In [10]:
import pandas as pd
import numpy as np
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import PorterStemmer, WordNetLemmatizer
import string
import re

nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')
nltk.download('omw-1.4')
nltk.download('punkt_tab')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\maila\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\maila\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\maila\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to
[nltk_data]     C:\Users\maila\AppData\Roaming\nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!
[nltk_data] Downloading package punkt_tab to
[nltk_data]     C:\Users\maila\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


True

### **2. Dataset Loading and Cleaning**

In [11]:
# Load the dataset
df = pd.read_csv('spam.csv', encoding='latin-1')

# Drop unnecessary columns
df = df.dropna(axis=1, how='any')

# Rename columns
df.columns = ['label', 'text']

df.head()

,label,text
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."


### **3. Implementation of NLP Preprocessing Techniques**

In [12]:
stop_words = set(stopwords.words('english'))
ps = PorterStemmer()
lemmatizer = WordNetLemmatizer()

def preprocess_text(text):
    # Lowercase and remove punctuation
    text = text.lower()
    text = re.sub(f"[{re.escape(string.punctuation)}]", "", text)
    
    # Tokenization
    tokens = word_tokenize(text)
    
    # Stopword Removal, Stemming and Lemmatization
    cleaned_tokens = [lemmatizer.lemmatize(ps.stem(w)) for w in tokens if w not in stop_words]
    
    return " ".join(cleaned_tokens)

df['processed_text'] = df['text'].apply(preprocess_text)
df[['text', 'processed_text']].head()

,text,processed_text
0,"Go until jurong point, crazy.. Available only ...",go jurong point crazi avail bugi n great world...
1,Ok lar... Joking wif u oni...,ok lar joke wif u oni
2,Free entry in 2 a wkly comp to win FA Cup fina...,free entri 2 wkli comp win fa cup final tkt 21...
3,U dun say so early hor... U c already then say...,u dun say earli hor u c alreadi say
4,"Nah I don't think he goes to usf, he lives aro...",nah dont think goe usf live around though


### **4. Text Vectorization**
- CountVectorizer
- TF-IDF

In [13]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

# TF-IDF Vectorization
tfidf = TfidfVectorizer(max_features=5000)
X = tfidf.fit_transform(df['processed_text']).toarray()
y = df['label'].map({'ham': 0, 'spam': 1})

print(f"Shape of X: {X.shape}")

Shape of X: (5572, 5000)


### **5. Build Machine Learning Classification Model**
- Train-Test Split (70-30)
- Model Training (Multinomial Naive Bayes)

In [14]:
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import classification_report, accuracy_score, confusion_matrix, precision_score, recall_score, f1_score

# Split the data (70% Train, 30% Test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=42)

# Train the model
model = MultinomialNB()
model.fit(X_train, y_train)

MultinomialNB()

### **6. Evaluate Model Performance**

In [15]:
y_pred = model.predict(X_test)

print("Accuracy Score:", accuracy_score(y_test, y_pred))
print("Precision Score:", precision_score(y_test, y_pred))
print("Recall Score:", recall_score(y_test, y_pred))
print("F1 Score:", f1_score(y_test, y_pred))

print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))

Accuracy Score: 0.9671052631578947
Precision Score: 1.0
Recall Score: 0.7488584474885844
F1 Score: 0.856396866840731

Confusion Matrix:
 [[1453    0]
 [  55  164]]

Classification Report:
               precision    recall  f1-score   support

           0       0.96      1.00      0.98      1453
           1       1.00      0.75      0.86       219

    accuracy                           0.97      1672
   macro avg       0.98      0.87      0.92      1672
weighted avg       0.97      0.97      0.97      1672



### **Conclusion**
In this assignment, we successfully implemented a complete NLP pipeline and built a text classification model:
1. **NLP Preprocessing**: We applied tokenization, stopword removal, stemming, and lemmatization to clean the raw text data.
2. **Text Vectorization**: We used **TF-IDF** to convert the processed text into numerical features suitable for machine learning.
3. **Model Building**: We implemented a **Multinomial Naive Bayes** classifier with a **70/30 train-test split**.
4. **Results**: The model achieved an accuracy of approximately **96.7%**, Recall of **74.88%**, with perfect precision for the spam class, indicating high reliability in classifying messages correctly.